src/typeahead/typeahead.ts

```diff
@@ -58,6 +58,42 @@ export interface NgbTypeaheadSelectItemEvent {
   preventDefault: () => void;
 }
 
+/**
+ * Extends standard Event interface to specify the type of the "target" property
+ * for the case of HTMLInputElement events.
+ */
+export interface HTMLInputElementEvent extends Event { target: HTMLInputElement & EventTarget; }
+
+/**
+ * Parameters given to the Typeahead user initialization function.
+ */
+export interface NgbTypeaheadInitParams {
+  /**
+   * Observable of input's input event, sending the event itself.
+   */
+  input$: Observable<HTMLInputElementEvent>;
+
+  /**
+   * Observable of input's focus event, sending the event itself.
+   */
+  focus$: Observable<HTMLInputElementEvent>;
+
+  /**
+   * Observable of input's click event, sending the event itself.
+   */
+  click$: Observable<HTMLInputElementEvent>;
+
+  /**
+   * The current NgbTypeahead instance.
+   */
+  instance: NgbTypeahead;
+
+  /**
+   * The custom context passed to this NgbTypeahead instance.
+   */
+  context: any;
```

Comment:
Not sure why we need it for focus handling

Reply:
For this feature, we don't need it, it was just a nice addition along with the others whose usefulness - I agree with you - can be discussed.

This was more specifically requested by @divdavem for a valid use case (which can be implemented without this but would be far more complex AFAIU).